add http2-debug-state feature

i110 opened the pull request:
Implemented `HTTP/2 Implementation Debug State` which is described below:
https://tools.ietf.org/html/draft-benfield-http2-debug-state-00

Sample Response:
```
< HTTP/2 200
< server: h2o/2.1.0-DEV
< date: Tue, 09 Aug 2016 09:50:17 GMT
< content-type: application/json; charset=utf-8
< cache-control: no-cache, no-store
< conn-flow-in: 65535
< conn-flow-out: 65535
<
{
  "settings": {
    "SETTINGS_HEADER_TABLE_SIZE": 4096,
    "SETTINGS_ENABLE_PUSH": 0,
    "SETTINGS_MAX_CONCURRENT_STREAMS": 100,
    "SETTINGS_INITIAL_WINDOW_SIZE": 16777216,
    "SETTINGS_MAX_FRAME_SIZE": 16384
  },
  "peerSettings": {
    "SETTINGS_HEADER_TABLE_SIZE": 4096,
    "SETTINGS_ENABLE_PUSH": 1,
    "SETTINGS_MAX_CONCURRENT_STREAMS": 4294967295,
    "SETTINGS_INITIAL_WINDOW_SIZE": 65535,
    "SETTINGS_MAX_FRAME_SIZE": 16384
  },
  "connFlowOut": 65535,
  "connFlowIn": 65535,
  "streams": {
    "1": {
      "state": "OPEN",
      "flowIn": 16777216,
      "flowOut": 65535,
      "dataIn": 0,
      "dataOut": 0
    }
  },
  "hpack": {
    "inbound_table_size": 3,
    "inbound_dynamic_header_table": [
      [ "accept", "*/*" ],
      [ "user-agent", "curl/7.49.1" ],
      [ ":authority", "127.0.0.1:8080" ]
    ],
    "outbound_table_size": 0,
    "outbound_dynamic_header_table": [
    ]
  },
  "sentGoAway": false
}
```

Points to consider:

* Using debugging URI with proxies may have security issues (being discussed at https://github.com/python-hyper/draft-http2-debug-state/issues/7)
    * When a connection sharing happens in a proxy, one's hpack state (including security-concerned fields) may be exposed to another
    * Perhaps the next draft will mention a switch which can be used to specify whether `hpack` property is included in the response or not (see: https://github.com/python-hyper/draft-http2-debug-state/commit/532c5315c8b8ab28993eff69e232dd261096359d)
    * This implementation doesn't have such a switch yet, because I understand and assume that in almost cases this feature will be used in the controllable environment, but we should watch and see the discussion, and add that switch when the next draft is published
* When a request comes via HTTP/1, I guess that the server should resnpond with 404
    * https://github.com/python-hyper/draft-http2-debug-state/issues/7

i110 commented on lib/handler/status.c:
```diff
@@ -112,7 +112,7 @@ static void send_response(struct st_h2o_status_collector_t *collector)
 
     req->res.status = 200;
-    h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, H2O_STRLIT("text/plain; charset=utf-8"));
+    h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_TYPE, H2O_STRLIT("application/json; charset=utf-8"));
```
@kazuho I thought this is better, WDYT?

kazuho replied:
Is this change relevant to the PR?
(and I believe we should use `text/plain` considering the fact that it can be displayed using web browsers...)

i110 replied:
Sorry, this isn't relevant to this PR.
I believe that canonical MIME type of JSON document is `application/json`, and also this json document are consumed by only "status" web-app, so people don't see in browsers, isn't it?
https://tools.ietf.org/html/rfc7159#section-11

But, I have no strong kodawari about this, so can get it back to`text/plain` ( also with `debug_state.c` )

i110 commented on include/h2o/http2_internal.h:
```diff
@@ -50,6 +50,12 @@ typedef struct st_h2o_hpack_header_table_t {
+    h2o_iovec_t *name;
+    h2o_iovec_t *value;
+    const char *err_desc; /* the recorded soft error description */
+} h2o_hpack_header_table_entry_t;
```
moved from `hpack.c` to expose it so that debug_state handler can use this..

i110 commented on lib/core/token_table.h:
```diff
@@ -21,68 +21,72 @@
  */
 
 /* DO NOT EDIT! generated by tokens.pl */
-h2o_token_t h2o__tokens[] = {{{H2O_STRLIT(":authority")}, 1, 0, 0, 0, 0},
```
doesn't change except for adding `conn-flow-in` and `conn-flow-out`, but re-formatted by `tokens.pl`

kazuho commented on lib/http2/hpack.c:
```diff
@@ -264,6 +258,11 @@ static inline struct st_h2o_hpack_header_table_entry_t *header_table_get(h2o_hpa
+{
+    return (h2o_hpack_header_table_entry_t*)(header_table_get(header_table, index));
+}
+
```
All `inline` functions must be also defined as `static`.

Also, I would suggest refraining from creating bridges (i.e. a function that just calls another function), since it makes the code-base unnecessary complex. Our general preference goes to keeping the code-base compact and precise, rather than minimizing impact of each PR.

In this case, I would suggest changing `static inline struct st_h2o_hpack_header_table_entry_t *header_table_get(...)` to `h2o_hpack_eader_table_entry_t *h2o_hpack_header_table_get(...)`. Even without `inline` attribute, compilers are allowed to (and will) inline-expand functions defined in the same source file as the caller.

i110 replied:
> keeping the code-base compact and precise, rather than minimizing impact of each PR
totally agree.
I'll fix.

kazuho commented:
Thank you for the PR! It's good to have the debugging feature on H2O.

Please take a look at my in-line comments as well as two high-level concerns stated below regarding the implementation.

Let’s not make `conn-flow-in` and `conn-flow-out` headers as tokens, considering the fact that adding a token has _some_ impact on performance (due to the increased lookup cost), and that the two headers would never be used in a performance-sensitive way.

How about defining the JSON generator as a connection-level callback (i.e. as one of the various h2-only callbacks in `h2o_conn_callbacks_t`. That way, there would be no need to expose the internals of HTTP/2 implementation to the handler.

kazuho commented on lib/handler/debug_state.c:
```diff
@@ -0,0 +1,236 @@
+
+    /* encode streams */
+    {
+        int coma_removed = 0;
```
s/coma/comma/g?

i110 replied:
I found this spelling in `lib/handler/status.c`, and misunderstood it as a oshare-way of comma.
I'll fix this!

kazuho commented on lib/handler/debug_state.c:
```diff
@@ -0,0 +1,236 @@
+                   stream->input_window._avail,
+                   stream->output_window._avail,
+                   (stream->_req_body == NULL ? 0 : stream->_req_body->size),
+                   stream->req.bytes_sent);
```
I would appreciate it if we could use memory less aggressively. In the current implementation, `OUTPUT` allocates 2KB chunk, and that means that we would consume at max 2*(100+100)=400KB of memory here.

i110 replied:
OK, I'll reduce memory footprint

kazuho commented on lib/handler/debug_state.c:
```diff
@@ -0,0 +1,236 @@
+                   ",\n"                                                                         \
+                   "      [ \"%s\", \"%s\" ]",                                                    \
+                   entry->name->base,                                                            \
+                   entry->value->base);                                                          \
```
Please use `%.*s` and specify the lengths of the values using `->len` for both `name` and `value`.

Also, please consider the case in which the name (or value) is very long. Current approach of allocating a fixed amount of memory would not work well.

i110 replied:
tashikani..

i110 commented:
@kazuho Thank you for your comments, I'll fix these problems,
BTW, new draft was published yesterday:
https://tools.ietf.org/html/draft-benfield-http2-debug-state-01
I'll follow this new draft from now.

kazuho commented:
Regarding memory footprint, since this is not a performance critical patch, I would suggest calling `snprintf` repeatedly until it succeeds in buiding the entire string, every time increasing the allocated buffer size by 2x (i.e. exponential back-off).

By using exponential back-off, you can guarantee that the size of memory consumed would be below the actual size of the output multiplied by a constant factor (which would be 3 in this case).

Note: from `man snprintf` on Mac OS X:
```
    The snprintf() and vsnprintf() functions will write at most size-1 of the characters printed into the output string
     (the size'th character then gets the terminating `\0'); if the return value is greater than or equal to the size
     argument, the string was too short and some of the printed characters were discarded.  The output is always null-ter-
     minated, unless size is 0.
```

kazuho commented:
Oops. Updated my previous comment. Exponential back-off isn't required in this case.

i110 commented:
@kazuho fixed issues.
Also cached up on draft-01. But some new added fields (`queuedData`, `paddingIn`, `paddingOut` ) are ommited because these are optional, and to add these fields we have to track additional states of streams, which I concerned about performance effect on.

kazuho commented on lib/handler/debug_state.c:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    const h2o_token_t conn_flow_in_token = { { H2O_STRLIT("conn-flow-in") } };
+    const h2o_token_t conn_flow_out_token = { { H2O_STRLIT("conn-flow-out") } };
```
You cannot do this.

`h2o_token_t` is an object that be compared by checking their address. To achieve such property, a string contained by a token MUST be interned (i.e. only one `h2o_token_t` should exist within the entire program for each string). Trying to define a token in each source file collides with the requirement.

You should instead use `h2o_add_header_by_str` for registering a header that is (either guaranteed or might) not be a token.

kazuho commented on include/h2o/http2.h:
```diff
@@ -60,6 +60,15 @@ extern const h2o_http2_priority_t h2o_http2_default_priority;
+typedef struct st_h2o_http2_debug_state_t{
+    h2o_iovec_vector_t json;
+    ssize_t conn_flow_in;
+    ssize_t conn_flow_out;
```
Use of `size_t` is preferred unless you need to use negative values to represent error conditions.

kazuho commented on include/h2o/http2_internal.h:
```diff
@@ -224,6 +232,7 @@ struct st_h2o_http2_conn_t {
     } _write;
     h2o_cache_t *push_memo;
     h2o_http2_casper_t *casper;
+    int _sent_goaway;
```
I believe `conn->state == H2O_HTTP2_CONN_STATE_HALF_CLOSED` would be sufficient to see if a GOAWAY frame has been sent. Please refer to the comments on `h2o_http2_conn_state_t`.

kazuho commented on lib/http2/debug_state.c:
```diff
@@ -0,0 +1,210 @@
+        v.base = h2o_mem_alloc_pool(&req->pool, size + 1);             \
+        v.len = snprintf(v.base, size + 1, _fmt, __VA_ARGS__);   \
+        state->json.entries[state->json.size++] = v; \
+    } while (0)
```
I'd suggest making this a function (instead of macro), considering the fact that the arguments are evaluated more than once, and because this is not performance-critical.

kazuho commented on lib/http2/debug_state.c:
```diff
@@ -0,0 +1,210 @@
+                comma_removed = 1;                                                               \
+            }                                                                                    \
+        }                                                                                        \
+    } while (0)
```
Let's also make this a function, so that it would be easier to maintain.

kazuho commented on lib/http2/debug_state.c:
```diff
@@ -0,0 +1,210 @@
+        case H2O_HTTP2_STREAM_STATE_IDLE:
+        case H2O_HTTP2_STREAM_STATE_END_STREAM:
+            return NULL;
+        }
```
State of a pulled stream changes to `half-closed (remote)` as soon as a HEADER or CONTINUATION frame with the `END_STREAM` flag being set is received by the server.

kazuho commented:
@i110 Thank you for the changes as well as following the updates to the I-D!

> * I couldn't found the benefit of using connection callbacks for JSON generation, so made the handler directly call `h2o_get_http2_debug_state` 

`if (req->version < 0x200) return NULL;` is not a type-safe way to check _which_ protocol implementation is used. Consider the case when HTTP/3 is introduced. Would this code work as expected? Using callback slots that gets initialized by each protocol handler does not have this issue (in other words, use of callbacks is type-safe).

And the added bonus would be that you can remove the tight coupling between lib/handler/debug_state.c and h2o/http2.h.

> * `s/coma/comma/g` also in `status.c` (If you hope another pr, I'll create it)
> * `application/json` についてはどうしますかね。 Should I restore it to `text/plain` ? https://github.com/h2o/h2o/pull/1019#discussion_r74546907

You should always create a separate PR for separate issues.

Consider the case when we find a bug related to this PR. If this PR was dedicated to the issue, then we can simply run `git revert`. Otherwise, we would need to look into each line of the code to see how to do it.

i110 commented:
@kazuho Thank you for your comments.
I fixed the issues you said, please take a look at commit comments.
Additionaly, as we talked, I removed `sentGoAway` field because of state tracking issue (and of cause this is optional)

kazuho commented on include/h2o.h:
```diff
@@ -673,6 +673,18 @@ typedef struct st_h2o_res_t {
+    ssize_t conn_flow_out;
+} h2o_debug_state_t;
+
+h2o_debug_state_t *h2o_http2_get_debug_state(h2o_req_t *req, int hpack_enabled);
```
Please move the prototype to http2_internal.h, since the function is going to be referred via callback from outside of the http2 stack.

kazuho commented on include/h2o.h:
```diff
@@ -673,6 +673,18 @@ typedef struct st_h2o_res_t {
+    h2o_iovec_vector_t json;
+    ssize_t conn_flow_in;
+    ssize_t conn_flow_out;
+} h2o_debug_state_t;
```
Please rename the struct back to `h2o_http2_debug_state_t`, since the structure is dependent to HTTP/2 (as per defined by the I-D, right?).

kazuho commented on lib/core/token_table.h:
```diff
@@ -21,67 +21,69 @@
+    { { H2O_STRLIT("x-forwarded-for") }, 0, 0, 0, 0, 0 },
+    { { H2O_STRLIT("x-reproxy-url") }, 0, 0, 0, 0, 0 },
+    { { H2O_STRLIT("x-traffic") }, 0, 0, 0, 0, 0 }
+};
```
Please completely revert the changes made to lib/core/token_table.h, so that the file would be left unchanged upon PR merger.

kazuho commented on lib/handler/configurator/debug_state.c:
```diff
@@ -0,0 +1,68 @@
+    default:
+        h2o_configurator_errprintf(cmd, node, "node must be a scalar or a mapping");
+        return -1;
+    }
```
How about defining this as a configuration directive that takes a scalar argument of `minimum`, `hpack`? Doing that way, we can eliminate the complexity of handling a mapping.

The handler would be _off_ in case the the configuration directive is not used.

i110 replied:
great idea!

kazuho commented on lib/handler/debug_state.c:
```diff
@@ -0,0 +1,72 @@
+        h2o_start_response(req, &generator);
+        h2o_send(req, NULL, 0, 1);
+        return 0;
+    }
```
Simply returning -1 (so that the next generator would handle the request) might be sufficient.

i110 replied:
tashikani.

kazuho commented on lib/http2/debug_state.c:
```diff
@@ -0,0 +1,217 @@
+    return NULL;
+}
+
+static void append_line(h2o_mem_pool_t *pool, h2o_iovec_vector_t *lines, const char *fmt, ...)
```
Please use ` __attribute__((format(printf, xxx, yyy)))` extension to detect errors in format specifiers. Examples can be found in the headers of H2O, and I assume documentation can be found on gcc.gnu.org.

kazuho commented on lib/http2/debug_state.c:
```diff
@@ -0,0 +1,217 @@
+    h2o_iovec_t v;
+    v.base = h2o_mem_alloc_pool(pool, size + 1);
+
+    va_start(args, fmt);
```
Please use `va_copy`. Applying `va_start` more than once against one `va_list` is not permitted (IIRC the behavior is undefined).

i110 replied:
memo:
As we talked, `va_start` can be called again after `va_end` is called against the same va_list

kazuho commented on lib/http2/debug_state.c:
```diff
@@ -0,0 +1,217 @@
+    if (hpack_enabled) {
+        nr_resp += 3 + conn->_input_header_table.num_entries + conn->_output_header_table.num_entries; // hpack
+    }
+    h2o_vector_reserve(&req->pool, &state->json, nr_resp);
```
Let's call h2o_vector_reserve in `append_line` instead of trying to determine the correct length of the output here.

This is not a performance sensitive handler, and minimizing the risks of having issues should be given higher priority than optimization.

kazuho commented:
@i110 Thank you for the updates.

I like the code. It has become even more precise and simple, it's getting near to merger.

Please take a look at my comments in-line.

i110 commented:
@kazuho: issues fixed :smile:

kazuho commented on lib/http2/debug_state.c:
```diff
@@ -0,0 +1,210 @@
+static const char *debug_state_string_half_closed_remote = "HALF_CLOSED_REMOTE";
+static const char *debug_state_string_reserved_local = "RESERVED_LOCAL";
+
+const char *get_debug_state_string(h2o_http2_stream_t *stream)
```
maybe add `static`?

kazuho commented on lib/http2/debug_state.c:
```diff
@@ -0,0 +1,210 @@
+    return NULL;
+}
+
+__attribute__((format(printf, 3, 4))) static void append_line(h2o_mem_pool_t *pool, h2o_iovec_vector_t *lines, const char *fmt, ...)
```
Renaming this function to `append_block` or `append_chunk` might be a good idea considering the fact that this function is never used to append a _line_.

kazuho commented:
Thank you for adding `http2`-prefix to the source file. Could you please do the same for the name of the configuration directive and for the global function exported by the handler?

kazuho commented on lib/handler/http2_debug_state.c:
```diff
@@ -0,0 +1,67 @@
+ */
+#include "h2o.h"
+
+#include <inttypes.h>
```
Please include system headers first and then project-specific headers.

This is necessary since we _might_ want to override some system-defined headers.

kazuho commented on lib/handler/http2_debug_state.c:
```diff
@@ -19,18 +19,17 @@
-struct st_h2o_debug_state_handler_t {
+struct st_h2o_http2_debug_state_handler_t {
     h2o_handler_t super;
     int hpack_enabled : 1;
```
If you want a boolean, please use `unsigned : 1` that takes either 0 or 1. `int :1` takes two either 0 or -1.

Or simply using `int` or `unsigned` without `:nnn` would be sufficient in this case, considering the fact that the size of the struct would be indifferent and that we wouldn't have many instances of this struct.

kazuho commented on lib/http2/http2_debug_state.c:
```diff
@@ -0,0 +1,212 @@
+                    "      [ \"%.*s\", \"%.*s\" ]",
+                    (int)entry->name->len, entry->name->base,
+                    (int)entry->value->len, entry->value->base);
+        if (!comma_removed) {
```
I assume you can change `!comma_removed` to `i == 0`?

i110 replied:
you're right..

kazuho commented on lib/http2/http2_debug_state.c:
```diff
@@ -0,0 +1,212 @@
+                    (int)entry->name->len, entry->name->base,
+                    (int)entry->value->len, entry->value->base);
+        if (!comma_removed) {
+            chunks->entries[chunks->size - 1].base[0] = ' ';
```
It might be preferable to do the following instead so as not to leave trailing spaces.
```
chunks->entries[chunk->size - 1].base += 1;
chunk->entries[chunk->size - 1].len -= 1;
```

It _might_ be better to define the logic as a separate function, since you would want to do the same at other parts of the code as well.

kazuho replied:
P.S. the other method you can use is to have comma appended to every element and then remove the last comma, or rewrite the first comma to { or [.

kazuho commented on lib/http2/hpack.c:
```diff
@@ -256,7 +250,7 @@ static h2o_iovec_t *decode_string(h2o_mem_pool_t *pool, const uint8_t **src, con
 }
 
-static inline struct st_h2o_hpack_header_table_entry_t *header_table_get(h2o_hpack_header_table_t *table, size_t index)
+h2o_hpack_header_table_entry_t *h2o_hpack_header_table_get(h2o_hpack_header_table_t *table, size_t index)
```
Assuming that all the structs used inside the function is defined in the headers, would it be possible to change this function to a static inline function declared and defined in http2_internal.h?

Please refer to other inline functions defined in http2_internal.h to how to define a global function as inline (in essence, declare as `static` and, define as `inline`).

kazuho commented on lib/http2/http2_debug_state.c:
```diff
@@ -86,19 +86,19 @@ __attribute__((format(printf, 3, 4))) static void append_chunk(h2o_mem_pool_t *p
+
+    if (i > 0)
+        // remove the last commna
+        --chunks->entries[chunks->size - 1].len;
```
Please do not insert a comment taking a line within a if statement (without using a block), since it is counterintuitive.

You should write the comment either _above_ the `if` keyword or surround the comment and the statement that is executed if `if` evaluates to true using `{}`.

i110 commented:
@kazuho fixed and applied clang-format.

kazuho commented on include/h2o.h:
```diff
@@ -1722,6 +1735,17 @@ void h2o_status_register(h2o_pathconf_t *pathconf);
  */
 void h2o_status_register_configurator(h2o_globalconf_t *conf);
 
+/* lib/handler/status.c */
```
wrong filename? :-)

i110 replied:
ぐぬぬ

kazuho commented on lib/http2/http2_debug_state.c:
```diff
@@ -0,0 +1,195 @@
+
+__attribute__((format(printf, 3, 4))) static void append_chunk(h2o_mem_pool_t *pool, h2o_iovec_vector_t *chunks, const char *fmt,
+                                                               ...)
+
```
unnecessary blank line here?

kazuho commented:
Going through the code I have found tiny tiny issues. Aside from that all look good to me.

i110 commented:
@kazuho fixed!

kazuho commented:
Thank you for your hard work!
